src/test/util_string_tests.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/string.h>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace util;
+
+BOOST_AUTO_TEST_SUITE(util_string_tests)
+
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
```

Comment:
> I think that real code should be designed without taking tests into account (apart from being testable). Tests should always follow real code, not the other way round.

Making the function return a value allowed for C) dropped negatives (and less runtime testing). But I can try to provide a new diff bringing them back and without changing *string.h*. My main point remains adding the interlaced tinyformat tests to prove parity.

> The diff you shared also changes `util::hex_literals::detail::Hex`, which I don't really understand.

It's a consequence of clashes between `util::detail` namespaces. (Added note about that in a later edit).

> All relevant test coverage has been taken over, except for the `*` test coverage. But the `*` parsing is broken in the linter anyway, so I am not sure about the value of taking over test in a comment (dead code).

Fair.

Reply:
<details>
<summary>New <code>git diff -w</code></summary>

```diff
diff --git a/src/test/util_string_tests.cpp b/src/test/util_string_tests.cpp
index c6ee1ffed9..3fed361e7e 100644
--- a/src/test/util_string_tests.cpp
+++ b/src/test/util_string_tests.cpp
@@ -10,62 +10,140 @@ using namespace util;
 
 BOOST_AUTO_TEST_SUITE(util_string_tests)
 
-BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+// These are counted correctly, since tinyformat will require the provided number of args.
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_CorrectCounts)
 {
-    // Compile-time sanity checks
-    static_assert([] {
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
+    tfm::format("");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%");
+    tfm::format("%%");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s");
+    BOOST_CHECK_THROW(tfm::format("%s"), tfm::format_error);
+    tfm::format("%s", "foo");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%s");
+    tfm::format("%%s");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("s%%");
+    tfm::format("s%%");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%%%s");
+    BOOST_CHECK_THROW(tfm::format("%%%s"), tfm::format_error);
+    tfm::format("%%%s", "foo");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s%%");
+    BOOST_CHECK_THROW(tfm::format("%s%%"), tfm::format_error);
+    tfm::format("%s%%", "foo");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers(" 1$s");
+    tfm::format(" 1$s");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$s");
+    BOOST_CHECK_THROW(tfm::format("%1$s"), tfm::format_error);
+    tfm::format("%1$s", "foo");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$s%1$s");
+    BOOST_CHECK_THROW(tfm::format("%1$s%1$s"), tfm::format_error);
+    tfm::format("%1$s%1$s", "foo");
+
     ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%2$s");
+    BOOST_CHECK_THROW(tfm::format("%2$s"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$s", "foo"), tfm::format_error);
+    tfm::format("%2$s", "foo", "bar");
+
     ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%2$s 4$s %2$s");
-        ConstevalFormatString<129>::Detail_CheckNumFormatSpecifiers("%129$s 999$s %2$s");
+    BOOST_CHECK_THROW(tfm::format("%2$s 4$s %2$s"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$s 4$s %2$s", "foo"), tfm::format_error);
+    tfm::format("%2$s 4$s %2$s", "foo", "bar");
+
+    ConstevalFormatString<12>::Detail_CheckNumFormatSpecifiers("%12$s 99$s %2$s");
+    BOOST_CHECK_THROW(tfm::format("%12$s 99$s %2$s"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%12$s 99$s %2$s", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"), tfm::format_error);
+    tfm::format("%12$s 99$s %2$s", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%02d");
+    BOOST_CHECK_THROW(tfm::format("%02d"), tfm::format_error);
+    tfm::format("%02d", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%+2s");
-        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%+2s");
+    BOOST_CHECK_THROW(tfm::format("%+2s"), tfm::format_error);
+    tfm::format("%+2s", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%.6i");
+    BOOST_CHECK_THROW(tfm::format("%.6i"), tfm::format_error);
+    tfm::format("%.6i", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%5.2f");
-        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%#x");
+    BOOST_CHECK_THROW(tfm::format("%5.2f"), tfm::format_error);
+    tfm::format("%5.2f", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$5i");
+    BOOST_CHECK_THROW(tfm::format("%1$5i"), tfm::format_error);
+    tfm::format("%1$5i", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$-5i");
+    BOOST_CHECK_THROW(tfm::format("%1$-5i"), tfm::format_error);
+    tfm::format("%1$-5i", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$.5i");
-        // tinyformat accepts almost any "type" spec, even '%', or '_', or '\n'.
+    BOOST_CHECK_THROW(tfm::format("%1$.5i"), tfm::format_error);
+    tfm::format("%1$.5i", 1);
+
+    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%#x");
+    BOOST_CHECK_THROW(tfm::format("%#x"), tfm::format_error);
+    tfm::format("%#x", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%123%");
+    BOOST_CHECK_THROW(tfm::format("%123%"), tfm::format_error);
+    tfm::format("%123%", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%123%s");
+    BOOST_CHECK_THROW(tfm::format("%123%s"), tfm::format_error);
+    tfm::format("%123%s", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_");
+    BOOST_CHECK_THROW(tfm::format("%_"), tfm::format_error);
+    tfm::format("%_", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%\n");
-        return true; // All checks above compiled and passed
-    }());
-    static_assert([] {
+    BOOST_CHECK_THROW(tfm::format("%\n"), tfm::format_error);
+    tfm::format("%\n", 1);
+}
+
 // The `*` specifier behavior is unsupported and can lead to runtime
 // errors when used in a ConstevalFormatString. Please refer to the
 // note in the ConstevalFormatString docs.
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_IncorrectCounts)
+{
+    int a{}, b{}, c{};
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%*c");
+    BOOST_CHECK_THROW(tfm::format("%*c"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%*c", a), tfm::format_error);
+    tfm::format("%*c", a, b);
+
     ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%2$*3$d");
+    BOOST_CHECK_THROW(tfm::format("%2$*3$d"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$*3$d", a), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$*3$d", a, b), tfm::format_error);
+    tfm::format("%2$*3$d", a, b, c);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%.*f");
-        return true; // All checks above compiled and passed
-    }());
+    BOOST_CHECK_THROW(tfm::format("%.*f"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%.*f", a), tfm::format_error);
+    tfm::format("%.*f", a, b);
+}
 
-    // Negative checks at runtime
+// Executed at runtime as exceptions are not allowed at compile time.
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_NegativeChecks)
+{
     using ErrType = const char*;
 
     auto check_mix{[](const ErrType& str) { return str == std::string_view{"Format specifiers must be all positional or all non-positional!"}; }};
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s%1$s"), ErrType, check_mix);
 
-    auto check_num_spec{[](const ErrType& str) { return str == std::string_view{"Format specifier count must match the argument count!"}; }};
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%s"), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%s"), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%1$s"), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%1$s"), ErrType, check_num_spec);
-
     auto check_0_pos{[](const ErrType& str) { return str == std::string_view{"Positional format specifier must have position of at least 1"}; }};
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%$s"), ErrType, check_0_pos);
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%$"), ErrType, check_0_pos);
```

</details>

#### Changes compared to fa26462e95291652b4021d91b014655f678149e8

- Interlaced tinyformat calls to prove and document parity between compile time and runtime checks. (The purpose isn't to to test tinyformat). Now using the less verbose `BOOST_CHECK_THROW` for tinyformat instead of `BOOST_CHECK_EXCEPTION`.
- Breaking apart test function into multiple functions.
- Still dropping a few negative tests, since they are proven by positive tests, even if less straightforward without the free function.

<details>
<summary>

#### Runtime performance comparison (+0.3ms/+1.8% longer)

</summary>

fa26462e95291652b4021d91b014655f678149e8 + diff:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.8 ms ±   0.2 ms    [User: 11.4 ms, System: 6.1 ms]
  Range (min … max):    16.2 ms …  17.8 ms    156 runs
```

fa26462e95291652b4021d91b014655f678149e8:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.5 ms ±   0.2 ms    [User: 10.6 ms, System: 6.5 ms]
  Range (min … max):    16.0 ms …  17.4 ms    154 runs
```
</details>

Reply:
<details>
<summary>

#### Diff on top of fae8c25d07cf003df1470699df4ef475055bb885

</summary>

```diff
diff --git a/src/test/util_string_tests.cpp b/src/test/util_string_tests.cpp
index 4c61632fce..ed383e9dad 100644
--- a/src/test/util_string_tests.cpp
+++ b/src/test/util_string_tests.cpp
@@ -17,7 +17,53 @@ template <unsigned NumArgs>
 inline void PassFmt(util::ConstevalFormatString<NumArgs> fmt)
 {
     decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt); // This was run at compile-time, but run it again at run-time to avoid -Wunused.
+
+    // Prove parity with tinyformat
+    switch (NumArgs) {
+    case 0:
+        tfm::format(fmt.fmt);
+        break;
+    case 1:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        tfm::format(fmt.fmt, "foo");
+        break;
+    case 2:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, "foo"), tfm::format_error);
+        tfm::format(fmt.fmt, "foo", "bar");
+        break;
+    case 12:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"), tfm::format_error);
+        tfm::format(fmt.fmt, "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12");
+        break;
+    }
 }
+
+template <unsigned WrongNumArgs, unsigned CorrectArgs>
+inline void PassFmtIncorrect(util::ConstevalFormatString<WrongNumArgs> fmt)
+{
+    decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt); // This was run at compile-time, but run it again at run-time to avoid -Wunused.
+
+    // Disprove parity with tinyformat
+    static_assert(WrongNumArgs != CorrectArgs);
+    int a{}, b{}, c{};
+    switch (CorrectArgs) {
+    case 2:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, a), tfm::format_error);
+        tfm::format(fmt.fmt, a, b);
+        break;
+    case 3:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, a), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, a, b), tfm::format_error);
+        tfm::format(fmt.fmt, a, b, c);
+        break;
+    }
+}
+
+// Executed at runtime as exceptions are not allowed at compile time.
 template <unsigned WrongNumArgs>
 inline void FailFmtWithError(std::string_view wrong_fmt, std::string_view error)
 {
@@ -40,7 +86,7 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     PassFmt<1>("%1$s%1$s");
     PassFmt<2>("%2$s");
     PassFmt<2>("%2$s 4$s %2$s");
-    PassFmt<129>("%129$s 999$s %2$s");
+    PassFmt<12>("%12$s 999$s %2$s");
     PassFmt<1>("%02d");
     PassFmt<1>("%+2s");
     PassFmt<1>("%.6i");
@@ -58,9 +104,9 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     // The `*` specifier behavior is unsupported and can lead to runtime
     // errors when used in a ConstevalFormatString. Please refer to the
     // note in the ConstevalFormatString docs.
-    PassFmt<1>("%*c");
-    PassFmt<2>("%2$*3$d");
-    PassFmt<1>("%.*f");
+    PassFmtIncorrect<1, 2>("%*c");
+    PassFmtIncorrect<2, 3>("%2$*3$d");
+    PassFmtIncorrect<1, 2>("%.*f");
 
     auto check_mix{"Format specifiers must be all positional or all non-positional!"};
     FailFmtWithError<1>("%s%1$s", check_mix);
```
</details>

#### Changes compared to fae8c25d07cf003df1470699df4ef475055bb885

tinyformat calls to prove and document parity/non-parity between compile time and runtime checks. (The purpose isn't to to test tinyformat). (Using the less verbose `BOOST_CHECK_THROW` for tinyformat instead of `BOOST_CHECK_EXCEPTION`).

<details>
<summary>

#### Runtime performance comparison (+0.1ms/+0.6% longer)

</summary

fae8c25d07cf003df1470699df4ef475055bb885 + diff:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.8 ms ±   0.4 ms    [User: 10.9 ms, System: 6.6 ms]
  Range (min … max):    16.4 ms …  19.9 ms    154 runs
```

fae8c25d07cf003df1470699df4ef475055bb885:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.7 ms ±   0.3 ms    [User: 10.8 ms, System: 6.6 ms]
  Range (min … max):    16.2 ms …  17.8 ms    157 runs
```
</details>

Reply:
> > Still dropping a few negative tests, since they are proven by positive tests, even if less straightforward without the free function.
>
> I don't think this is true. Without the free function

True, forgot that that last condition was back in the non-free function. Negative tests are not removed in [latest diff](https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1756443830).

Reply:
> Otherwise, it will never be ready because there is always a way to come up with an idea to improve something

I think we're getting closer, your patience is appreciated :)

Reply:
Yeah, I can do it as a follow-up since by now it does not really modify your existing code that much.

If you need to re-touch, I would appreciate if you just took this diff:
```diff
-    PassFmt<129>("%129$s 999$s %2$s");
+    PassFmt<12>("%12$s 999$s %2$s");
```
to make the follow-up cleaner.